File: src/Persistence/Repositories/Accountant/TransactionsRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;
using Application.Contracts.Accountant.Transactions;
using Dapper;
using Domain.Entities.Accountant;
using Microsoft.EntityFrameworkCore;

namespace Persistence.Repositories.Accountant;

public class TransactionsRepository : BaseRepository, ITransactionsRepository
{
    public TransactionsRepository(PersonalAssistantContext efContext)
        : base(efContext) { }

    public IEnumerable<Transaction> GetAllForExport(int userId, string uncategorized)
    {
        using IDbConnection conn = OpenConnection();

        const string query = @"SELECT t.*, fa.id, fa.name, ta.id, ta.name, c.id, c.name, pc.id, pc.name
                        FROM accountant_transactions AS t
                        LEFT JOIN accountant_accounts AS fa ON t.from_account_id = fa.id
                        LEFT JOIN accountant_accounts AS ta ON t.to_account_id = ta.id
                        LEFT JOIN accountant_categories AS c ON t.category_id = c.id
                        LEFT JOIN accountant_categories AS pc ON c.parent_id = pc.id
                        WHERE fa.user_id = @UserId OR ta.user_id = @UserId ORDER BY date";

        var transactions = conn.Query<Transaction, Account, Account, Category, Category, Transaction>(query,
            (transaction, fromAccount, toAccount, category, parentCategory) =>
            {
                transaction.FromAccount = fromAccount ?? new Account();
                transaction.ToAccount = toAccount ?? new Account();

                if (category == null)
                {
                    transaction.Category = new Category { Name = uncategorized };
                }
                else
                {
                    if (parentCategory != null)
                    {
                        category.Name = $"{parentCategory.Name}/{category.Name}";
                    }

                    transaction.Category = category;
                }

                return transaction;
            }, new { UserId = userId });

        return transactions;
    }

    public IEnumerable<Transaction> GetAll(int userId, DateTime fromModifiedDate)
    {
        using IDbConnection conn = OpenConnection();

        return conn.Query<Transaction>(@"SELECT t.* 
                                         FROM accountant_transactions AS t
                                         INNER JOIN accountant_accounts AS a ON a.id = t.from_account_id 
                                             OR a.id = t.to_account_id 
                                         WHERE a.user_id = @UserId AND t.modified_date > @FromModifiedDate",
            new { UserId = userId, FromModifiedDate = fromModifiedDate });
    }

    public IEnumerable<int> GetDeletedIds(int userId, DateTime fromDate)
    {
        using IDbConnection conn = OpenConnection();

        return conn.Query<int>(@"SELECT entity_id FROM accountant_deleted_entities WHERE user_id = @UserId AND entity_type = @EntityType AND deleted_date > @DeletedDate",
            new { UserId = userId, EntityType = (short)EntityType.Transaction, DeletedDate = fromDate });
    }

    public async Task<int> CreateAsync(Transaction transaction)
    {
        using IDbConnection conn = OpenConnection();

        int id;
        var insertQuery = @"INSERT INTO accountant_transactions 
                                (from_account_id, to_account_id, category_id, amount, from_stocks, to_stocks, currency, description, date, is_encrypted, encrypted_description, salt, nonce, created_date, modified_date)
                                VALUES 
                                (@FromAccountId, @ToAccountId, @CategoryId, @Amount, @FromStocks, @ToStocks, @Currency, @Description, @Date, @IsEncrypted, @EncryptedDescription, @Salt, @Nonce, @CreatedDate, @ModifiedDate) returning id";

        var dbTransaction = conn.BeginTransaction();

        id = (await conn.QueryAsync<int>(insertQuery, transaction, dbTransaction)).Single();

        if (transaction.FromAccountId.HasValue && !transaction.ToAccountId.HasValue)
        {
            var relatedUpcomingExpenses = conn.Query<UpcomingExpense>(@"SELECT * FROM accountant_upcoming_expenses
                                                                        WHERE category_id = @CategoryId 
                                                                            AND EXTRACT(year FROM date) = @Year
                                                                            AND EXTRACT(month FROM date) = @Month",
                new { transaction.CategoryId, transaction.Date.Year, transaction.Date.Month }).ToList();

            if (relatedUpcomingExpenses.Any())
            {
                bool transactionHasDescription = !string.IsNullOrEmpty(transaction.Description);

                foreach (var upcomingExpense in relatedUpcomingExpenses)
                {
                    bool upcomingExpenseHasDescription = !string.IsNullOrEmpty(upcomingExpense.Description);
                    bool bothWithDescriptionsAndTheyMatch = upcomingExpenseHasDescription
                                                            && transactionHasDescription
                                                            && string.Equals(upcomingExpense.Description, transaction.Description.Trim(), StringComparison.OrdinalIgnoreCase);

                    if (!upcomingExpenseHasDescription || bothWithDescriptionsAndTheyMatch)
                    {
                        if (upcomingExpense.Amount > transaction.Amount)
                        {
                            await conn.ExecuteAsync(@"UPDATE accountant_upcoming_expenses SET amount = amount - @Amount, modified_date = @ModifiedDate WHERE id = @Id",
                                new { upcomingExpense.Id, transaction.Amount, ModifiedDate = DateTime.UtcNow }, dbTransaction);
                        }
                        else
                        {
                            await conn.ExecuteAsync(@"DELETE FROM accountant_upcoming_expenses WHERE id = @Id",
                                new { upcomingExpense.Id }, dbTransaction);

                            await conn.QueryAsync<int>(@"INSERT INTO accountant_deleted_entities (user_id, entity_type, entity_id, deleted_date)
                                                         VALUES (@UserId, @EntityType, @EntityId, @DeletedDate)",
                                new { upcomingExpense.UserId, EntityType = (short)EntityType.UpcomingExpense, EntityId = upcomingExpense.Id, DeletedDate = DateTime.UtcNow },
                                dbTransaction);
                        }
                    }
                }
            }
        }

        dbTransaction.Commit();

        return id;
    }

    public async Task UpdateAsync(Transaction transaction)
    {
        Transaction dbTransaction = EFContext.Transactions.Find(transaction.Id);

        dbTransaction.FromAccountId = transaction.FromAccountId;
        dbTransaction.ToAccountId = transaction.ToAccountId;
        dbTransaction.CategoryId = transaction.CategoryId;
        dbTransaction.Amount = transaction.Amount;
        dbTransaction.FromStocks = transaction.FromStocks;
        dbTransaction.ToStocks = transaction.ToStocks;
        dbTransaction.Currency = transaction.Currency;
        dbTransaction.Description = transaction.Description;
        dbTransaction.Date = transaction.Date.ToUniversalTime();
        dbTransaction.IsEncrypted = transaction.IsEncrypted;
        dbTransaction.EncryptedDescription = transaction.EncryptedDescription;
        dbTransaction.Salt = transaction.Salt;
        dbTransaction.Nonce = transaction.Nonce;
        dbTransaction.ModifiedDate = transaction.ModifiedDate;

        await EFContext.SaveChangesAsync();
    }

    public async Task DeleteAsync(int id, int userId)
    {
        var deletedEntity = EFContext.DeletedEntities.FirstOrDefault(x => x.UserId == userId && x.EntityType == EntityType.Transaction && x.EntityId == id);
        if (deletedEntity == null)
        {
            EFContext.DeletedEntities.Add(new DeletedEntity
            {
                UserId = userId,
                EntityType = EntityType.Transaction,
                EntityId = id,
                DeletedDate = DateTime.UtcNow
            });
        }
        else
        {
            deletedEntity.DeletedDate = DateTime.UtcNow;
        }

        var transaction = EFContext.Transactions.First(x => x.Id == id 
            && !x.FromAccountId.HasValue || x.FromAccount.UserId == userId
            && !x.ToAccountId.HasValue || x.ToAccount.UserId == userId);

        EFContext.Transactions.Remove(transaction);

        await EFContext.SaveChangesAsync();
    }
}

File: src/Infrastructure/Accountant.Persistence.Fs/TransactionsRepository.fs
namespace Accountant.Persistence.Fs

open System
open Npgsql.FSharp
open ConnectionUtils
open Models
open Npgsql

module TransactionsRepository =

    [<Literal>]
    let private table = "accountant.transactions"

    let getAll (userId: int) (fromModifiedDate: DateTime) (conn: RegularOrTransactionalConn) =
        ConnectionUtils.connect conn
        |> Sql.query
            $"SELECT t.* 
                      FROM {table} AS t
                      INNER JOIN accountant.accounts AS a ON a.id = t.from_account_id 
                          OR a.id = t.to_account_id 
                      WHERE a.user_id = @user_id AND t.modified_date > @modified_date"
        |> Sql.parameters [ "user_id", Sql.int userId; "modified_date", Sql.timestamptz fromModifiedDate ]
        |> Sql.executeAsync (fun read ->
            { Id = read.int "id"
              FromAccountId = read.intOrNone "from_account_id"
              ToAccountId = read.intOrNone "to_account_id"
              CategoryId = read.intOrNone "category_id"
              Amount = read.decimal "amount"
              FromStocks = read.decimalOrNone "from_stocks"
              ToStocks = read.decimalOrNone "to_stocks"
              Currency = read.string "currency"
              Description = read.stringOrNone "description"
              Date = read.dateTime "date"
              IsEncrypted = read.bool "is_encrypted"
              EncryptedDescription = read.byteaOrNone "encrypted_description"
              Salt = read.byteaOrNone "salt"
              Nonce = read.byteaOrNone "nonce"
              Generated = read.bool "generated"
              CreatedDate = read.dateTime "created_date"
              ModifiedDate = read.dateTime "modified_date" })

    let exists (id: int) (userId: int) (conn: RegularOrTransactionalConn) =
        ConnectionUtils.connect conn
        |> Sql.query $"SELECT COUNT(*) AS count
                       FROM {table} AS t
                       LEFT JOIN accountant.accounts AS fa ON t.from_account_id = fa.id
                       LEFT JOIN accountant.accounts AS ta ON t.to_account_id = ta.id
                       WHERE t.id = @id AND (t.from_account_id IS NULL OR fa.user_id = @user_id) AND (t.to_account_id IS NULL OR ta.user_id = @user_id)"
        |> Sql.parameters [ "id", Sql.int id; "user_id", Sql.int userId ]
        |> Sql.executeRow (fun read -> (read.int "count") > 0)

    let create (transaction: Transaction) (conn: RegularOrTransactionalConn) (tran: NpgsqlTransaction Option) =
        let npgsqlConn =
            match conn with
            | ConnectionString cs ->
                let conn = new NpgsqlConnection(cs)
                conn.Open()
                conn
            | TransactionConnection tc -> tc

        // If there is an outside transaction use that one
        // Otherwise begin a new one
        let tr =
            match tran with
            | Some t -> t
            | None -> npgsqlConn.BeginTransaction()

        task {
            let! id =
                npgsqlConn
                |> Sql.existingConnection
                |> Sql.query
                    $"INSERT INTO {table}
                            (from_account_id, to_account_id, category_id, amount, from_stocks, to_stocks, currency, description, date, is_encrypted, encrypted_description, salt, nonce, generated, created_date, modified_date) VALUES 
                            (@from_account_id, @to_account_id, @category_id, @amount, @from_stocks, @to_stocks, @currency, @description, @date, @is_encrypted, @encrypted_description, @salt, @nonce, @generated, @created_date, @modified_date) RETURNING id"
                |> Sql.parameters
                    [ "from_account_id", Sql.intOrNone transaction.FromAccountId
                      "to_account_id", Sql.intOrNone transaction.ToAccountId
                      "category_id", Sql.intOrNone transaction.CategoryId
                      "amount", Sql.decimal transaction.Amount
                      "from_stocks", Sql.decimalOrNone transaction.FromStocks
                      "to_stocks", Sql.decimalOrNone transaction.ToStocks
                      "currency", Sql.string transaction.Currency
                      "description", Sql.stringOrNone transaction.Description
                      "date", Sql.date (transaction.Date.ToUniversalTime())
                      "is_encrypted", Sql.bool transaction.IsEncrypted
                      "encrypted_description", Sql.byteaOrNone transaction.EncryptedDescription
                      "salt", Sql.byteaOrNone transaction.Salt
                      "nonce", Sql.byteaOrNone transaction.Nonce
                      "generated", Sql.bool transaction.Generated
                      "created_date", Sql.timestamptz transaction.CreatedDate
                      "modified_date", Sql.timestamptz transaction.ModifiedDate ]
                |> Sql.executeRowAsync (fun read -> read.int "id")
                |> Async.AwaitTask

            if transaction.FromAccountId.IsSome && transaction.ToAccountId.IsNone then
                let relatedUpcomingExpenses =
                    npgsqlConn
                    |> Sql.existingConnection
                    |> Sql.query
                        "SELECT * FROM accountant.upcoming_expenses
                            WHERE category_id = @category_id 
                                AND EXTRACT(year FROM date) = @year
                                AND EXTRACT(month FROM date) = @month"
                    |> Sql.parameters
                        [ "category_id", Sql.intOrNone transaction.CategoryId
                          "year", Sql.int (transaction.Date.Year)
                          "month", Sql.int (transaction.Date.Month) ]
                    |> Sql.execute (fun read ->
                        { Id = read.int "id"
                          UserId = read.int "user_id"
                          CategoryId = read.intOrNone "category_id"
                          Amount = read.decimal "amount"
                          Currency = read.string "currency"
                          Description = read.stringOrNone "description"
                          Date = read.dateTime "date"
                          Generated = read.bool "generated"
                          CreatedDate = read.dateTime "created_date"
                          ModifiedDate = read.dateTime "modified_date" })

                for ue in relatedUpcomingExpenses do
                    let bothWithDescriptionsAndTheyMatch =
                        ue.Description.IsSome
                        && transaction.Description.IsSome
                        && String.Equals(
                            ue.Description.Value,
                            transaction.Description.Value,
                            StringComparison.OrdinalIgnoreCase
                        )

                    if ue.Description.IsNone || bothWithDescriptionsAndTheyMatch then
                        if ue.Amount > transaction.Amount then
                            npgsqlConn
                            |> Sql.existingConnection
                            |> Sql.query
                                "UPDATE accountant.upcoming_expenses
                                    SET amount = amount - @Amount, modified_date = @ModifiedDate
                                    WHERE id = @Id"
                            |> Sql.parameters
                                [ "id", Sql.int ue.Id
                                  "amount", Sql.decimal transaction.Amount
                                  "modified_date", Sql.timestamptz DateTime.UtcNow ]
                            |> Sql.executeNonQueryAsync
                            |> Async.AwaitTask
                            |> ignore
                        else
                            npgsqlConn
                            |> Sql.existingConnection
                            |> Sql.executeTransactionAsync
                                [ "INSERT INTO accountant.deleted_entities
                                      (user_id, entity_type, entity_id, deleted_date) VALUES
                                      (@user_id, @entity_type, @entity_id, @deleted_date)",
                                  [ [ "user_id", Sql.int ue.UserId
                                      "entity_type", Sql.int (LanguagePrimitives.EnumToValue EntityType.UpcomingExpense)
                                      "entity_id", Sql.int id
                                      "deleted_date", Sql.timestamptz DateTime.UtcNow ] ]

                                  $"DELETE FROM {table} WHERE id = @id AND user_id = @user_id",
                                  [ [ "id", Sql.int id; "user_id", Sql.int ue.UserId ] ] ]
                            |> Async.AwaitTask
                            |> ignore

            if tran.IsNone then
                tr.Commit()

            return id
        }

    let update (transaction: Transaction) (conn: RegularOrTransactionalConn) =
        task {
            return!
                ConnectionUtils.connect conn
                |> Sql.query
                    $"UPDATE {table}
                           SET from_account_id = @from_account_id, to_account_id = @to_account_id, category_id = @category_id, amount = @amount, from_stocks = @from_stocks,
                                to_stocks = @to_stocks, currency = @currency, description = @description, date = @date, is_encrypted = @is_encrypted,
                                encrypted_description = @encrypted_description, salt = @salt, nonce = @nonce, modified_date = @modified_date
                           WHERE id = @id"
                |> Sql.parameters
                    [ "id", Sql.int transaction.Id
                      "from_account_id", Sql.intOrNone transaction.FromAccountId
                      "to_account_id", Sql.intOrNone transaction.ToAccountId
                      "category_id", Sql.intOrNone transaction.CategoryId
                      "amount", Sql.decimal transaction.Amount
                      "from_stocks", Sql.decimalOrNone transaction.FromStocks
                      "to_stocks", Sql.decimalOrNone transaction.ToStocks
                      "currency", Sql.string transaction.Currency
                      "description", Sql.stringOrNone transaction.Description
                      "date", Sql.date (transaction.Date.ToUniversalTime())
                      "is_encrypted", Sql.bool transaction.IsEncrypted
                      "encrypted_description", Sql.byteaOrNone transaction.EncryptedDescription
                      "salt", Sql.byteaOrNone transaction.Salt
                      "nonce", Sql.byteaOrNone transaction.Nonce
                      "modified_date", Sql.timestamptz transaction.ModifiedDate ]
                |> Sql.executeNonQueryAsync
                |> Async.AwaitTask
        }

    let delete (id: int) (userId: int) (conn: RegularOrTransactionalConn) =
        task {
            ConnectionUtils.connect conn
            |> Sql.executeTransactionAsync
                [ "INSERT INTO accountant.deleted_entities
                      (user_id, entity_type, entity_id, deleted_date) VALUES
                      (@user_id, @entity_type, @entity_id, @deleted_date)",
                  [ [ "user_id", Sql.int userId
                      "entity_type", Sql.int (LanguagePrimitives.EnumToValue EntityType.Transaction)
                      "entity_id", Sql.int id
                      "deleted_date", Sql.timestamptz DateTime.UtcNow ] ]

                  $"DELETE FROM accountant.transactions t
                        USING accountant.accounts a
                    WHERE t.id = @id AND 
                        ((t.from_account_id = a.id AND a.user_id = @user_id) OR (t.to_account_id = a.id AND a.user_id = @user_id))",
                  [ [ "id", Sql.int id; "user_id", Sql.int userId ] ] ]
            |> Async.AwaitTask
            |> ignore
        }
